source/common/upstream/load_balancer_impl.h

```diff
@@ -90,9 +91,24 @@ class LoadBalancerBase : public LoadBalancer {
     return per_priority_load_.degraded_priority_load_.get()[priority];
   }
   bool isInPanic(uint32_t priority) const { return per_priority_panic_[priority]; }
+  uint64_t random(bool peeking) {
+    if (peeking) {
+      stashed_random_.push_back(random_.random());
+      return stashed_random_.back();
+    } else {
+      if (!stashed_random_.empty()) {
+        auto random = stashed_random_.front();
+        stashed_random_.pop_front();
+        return random;
+      } else {
+        return random_.random();
+      }
+    }
+  }
 
   ClusterStats& stats_;
   Runtime::Loader& runtime_;
+  std::list<uint64_t> stashed_random_;
```

Comment:
perf nit: would a dequeue be better here? Maybe overkill for this use case but seems like we could avoid memory allocation/free in the common case.